README.md
---
languages:
- python
products:
- azure
- azure-machine-learning-service
description: "CLI for starting, stopping, sshing, and other ways of interacting with your Azure ML VM instance"
---

## MLOps with Azure ML

# To install
* Clone the repository
* cd into the repository
* `conda env create -f env.yml`
* `conda activate mlvm`


command line options

mlvm setup (python setup.py)
- runs mlvm conda, mlvm create
- After that is set up, you will configure the mlvm CLI tool to interact with your instance

mlvm conda
- create your mlvm conda environment (if not already created) and activate it

mlvm create
- You will be prompted to create your own azure ml vm which will be found here: https://ml.azure.com/compute/list/instances?wsid=/subscriptions/a1eab4f0-e17c-4e70-ab04-833c063dc515/resourceGroups/dt-datascience-core-dev-usw2/providers/Microsoft.MachineLearningServices/workspaces/dt-datascience-core-dev-usw2&tid=1aff0669-ee5f-40b8-9800-b5ec4f39c48e
- If you already have one, then just enter your instance name when prompted

mlvm start
- start your vm instance

mlvm stop
- stop your vm instance

mlvm ssh
- ssh into your vm instance


CHANGELOG.md
# Changelog

Azure ML VM interface

All notable changes to this project will be documented in this file.

## [1.0.0] - 2023-10-05
* 

## [0.1.0] - 2023-10-02
* Initial commit


tests/unit/data_test.py
# test integrity of the input dataBILITY OF SUCH DAMAGE.

import os
import numpy as np
import pandas as pd


# get absolute path of csv files from data folder
def get_absPath(filename):
    """Returns the path of the notebooks folder"""
    path = os.path.abspath(
        os.path.join(
            os.path.dirname(
                __file__), os.path.pardir, os.path.pardir, "data", filename
        )
    )
    return path


# number of features
expected_columns = 10

# distribution of features in the training set
historical_mean = np.array(
    [
        -3.63962254e-16,
        1.26972339e-16,
        -8.01646331e-16,
        1.28856202e-16,
        -8.99230414e-17,
        1.29609747e-16,
        -4.56397112e-16,
        3.87573332e-16,
        -3.84559152e-16,
        -3.39848813e-16,
        1.52133484e02,
    ]
)
historical_std = np.array(
    [
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        4.75651494e-02,
        7.70057459e01,
    ]
)

# maximal relative change in feature mean or standrd deviation
# that we can tolerate
shift_tolerance = 3


def test_check_schema():
    datafile = get_absPath("diabetes.csv")
    # check that file exists
    assert os.path.exists(datafile)
    dataset = pd.read_csv(datafile)
    header = dataset[dataset.columns[:-1]]
    actual_columns = header.shape[1]
    # check header has expected number of columns
    assert actual_columns == expected_columns


def test_check_bad_schema():
    datafile = get_absPath("diabetes_bad_schema.csv")
    # check that file exists
    assert os.path.exists(datafile)
    dataset = pd.read_csv(datafile)
    header = dataset[dataset.columns[:-1]]
    actual_columns = header.shape[1]
    # check header has expected number of columns
    assert actual_columns != expected_columns


def test_check_missing_values():
    datafile = get_absPath("diabetes_missing_values.csv")
    # check that file exists
    assert os.path.exists(datafile)
    dataset = pd.read_csv(datafile)
    n_nan = np.sum(np.isnan(dataset.values))
    assert n_nan > 0


def test_check_distribution():
    datafile = get_absPath("diabetes_bad_dist.csv")
    # check that file exists
    assert os.path.exists(datafile)
    dataset = pd.read_csv(datafile)
    mean = np.mean(dataset.values, axis=0)
    std = np.mean(dataset.values, axis=0)
    assert (
        np.sum(abs(mean - historical_mean) >
               shift_tolerance * abs(historical_mean))
        or np.sum(abs(std - historical_std) >
                  shift_tolerance * abs(historical_std)) > 0
    )


lib/log.py
import sys

# the verbose option indicates that this message will only appear in verbose mode
# verbose mode is defined to be true if '-v' or '--verbose' exists in sys.argv

# Recommended that level be one of INFO, WARN, ERROR, DEBUG
def disp_message(level, message, verbose):
    statement = f"{level}: {message}"
    # colorize the statement if warn or error level
    if level == 'WARN':
        statement = f"\033[93m{statement}\033[00m" # yellow
    elif level == 'ERROR':
        statement = f"\033[91m {statement}\033[00m" # red

    if not verbose:
        print(statement)
    elif is_verbose(): # if we are in verbose mode
        print(f'V {statement}')
    # else, it requires verbose mode to print but we're not in verbose mode

def info(message='', verbose=False):
    disp_message('INFO', message, verbose)

def warn(message='', verbose=False):
    disp_message('WARN', message, verbose)

def error(message='', verbose=False):
    disp_message('ERROR', message, verbose)
    exit(1)

# assumes we're in debug mode if sys.argv contains '--debug'
# else, do nothing
def debug(message=''):
    if is_debug:
        disp_message('DEBUG', message, verbose=False)

def usr_in(prompt=''):
    return input(prompt + ' ')

def yn(prompt=''):
    usr_in = input(prompt + ' (y/n) ')
    return usr_in in ('y', 'yes')

def is_verbose():
    return '-v' in sys.argv or '--verbose' in sys.argv

def is_debug():
    return '--debug' in sys.argv


src/cli.py
import os, re
from src.mlvm import Mlvm
from src import const

class Cli():
    def __init__(self):
        self.mlvm = Mlvm()

    # help='Display the current version of this CLI'
    def version(self):
        changelog_txt = open("CHANGELOG.md").read()
        version = re.findall(r"\[\d.*\].*", changelog_txt)[0]
        print(f"Azure ML VM CLI version {version}")

    # help="Create/Link your VM and set up this CLI to your path"
    # '--vm-name', help="If not set, you will be prompted to enter a vm name. Recommended format: ds-vm-<yourname>"
    def setup(self, workspace=const.DEFAULT_WORKSPACE, resource_group=const.DEFAULT_RESOURCE_GROUP, subscription_id=const.DEFAULT_SUBSCRIPTION_ID,
              vm_name=None, vm_size=const.DEFAULT_VM_SIZE, ssh_user=const.DEFAULT_SSH_USER,
              ssh_public_key_file=os.path.join(const.SSH_DIR,'id_rsa.pub'), ssh_private_key_file=os.path.join(const.SSH_DIR,'id_rsa')):
        return self.mlvm.setup(workspace, resource_group, subscription_id, vm_name, vm_size, ssh_user, ssh_public_key_file, ssh_private_key_file)

    # help="Pull the latest mlvm version, update the conda env, and generate new executable"
    def update(self):
        return self.mlvm.update()

    # help='Start up your Azure ML VM'
    # '-w', '--wait', help='Wait for the VM to become available')
    def start(self, wait=False):
        return self.mlvm.start(wait)

    # name='stop', help='Shut down your Azure ML VM'
    # '-w', '--wait', help='Wait for the VM to shut down')
    def stop(self, wait=False):
        return self.mlvm.stop(wait)

    # help='SSH into your Azure ML VM'
    def ssh(self):
        return self.mlvm.ssh()


lib/conda.py
# setting up conda environment, assumes conda is already installed

import subprocess
import re

class Conda:

    def __init__(self, env_name=None, env_file='env.yml'):
        if env_name is None:
            self.env_name = self.name_from_env_file(env_file)
        else:
            self.env_name = env_name
        self.env_file = env_file

    @staticmethod
    def name_from_env_file(env_file):
        # pyyaml is not a standard python library
        # so we have to parse the file the old fashioned way
        # We're looking for "name: <env_name>"
        with open(env_file, 'r') as f:
            for line in f.readlines():
                match = re.match(r'name\: (.*)', line)
                if match:
                    return match.group(1)

    def setup(self):
        # ensure conda is installed
        if not self.conda_installed():
            print('conda not installed. You can install Anaconda from https://www.anaconda.com/download')
            exit(1)

        if not self.env_exists():
            # we need to create the environment first
            self.create_env()
        else: # we need to ensure the environment is up to date
            self.update_env()
        self.activate_env()

    def env_exists(self):
        env_list = subprocess.check_output('conda env list'.split(' '), text=True)
        for line in env_list.split("\n"):
            if re.match(fr'^{self.env_name} ', line):
                return True
        return False

    def conda_installed(self):
        try:
            subprocess.check_call('conda --version'.split(' '))
            return True
        except FileNotFoundError:
            return False

    def create_env(self):
        subprocess.run(f'conda env create -f {self.env_file}'.split(' '))

    def update_env(self):
        subprocess.run(f'conda env update -f {self.env_file}'.split(' '))

    def activate_env(self):
        subprocess.run(f'conda activate {self.env_name}'.split(' '))

    def deactivate_env(self):
        subprocess.run(f'conda deactivate {self.env_name}'.split(' '))


src/mlvm.py
import string, os, shutil, socket, subprocess, json, paramiko, git
from cryptography.hazmat import primitives as crypto
import PyInstaller.__main__
from src import const
from lib import log
from lib.conda import Conda
from azure.core.exceptions import ResourceNotFoundError
from azure.identity import DefaultAzureCredential, InteractiveBrowserCredential
from azure.ai.ml import MLClient
from azure.ai.ml.entities import ComputeInstance, ComputeInstanceSshSettings


class Mlvm():

    def __init__(self):
        __class__.create_config_file_if_not_exists()
        self.azureml_config = self.load_azureml_config()
        self.ml_client = None # we'll create this as needed
        self.compute_instance = None # we'll set this as needed


    ## COMMAND IMPLEMENTATIONS

    def setup(self, workspace, resource_group, subscription_id, vm_name, vm_size, ssh_user, ssh_public_key_file, ssh_private_key_file):
        # validate the vm name and vm size
        if not vm_name is None and not __class__.validate_vm_name(vm_name):
            log.error("Invalid vm name. Must be between 2-16 chars and contain only letters, numbers, or -")
        if not __class__.validate_vm_size(vm_size): # vm_size should not be None. It has a default.
            log.error(f"Invalid vm size. It must be one of {const.VALID_VM_SIZES}")
        # populate the azureml config with the supplied arguments
        self.initialize_azureml_config(self, workspace, resource_group, subscription_id, vm_name, vm_size, ssh_user, ssh_public_key_file, ssh_private_key_file)

        self.set_ml_client()
        ssh_settings = self.setup_ssh()
        vm_linked = self.link_vm(ssh_settings) # either sets self.compute_instance and returns True or leaves it as None and returns False
        if not vm_linked:
            log.error("Didn't link VM. Setup aborted.")
        self.setup_cli() # create the cli executable add it to PATH
        self.setup_sf_connection()

    def update(self):
        log.info("Pulling latest version of mlvm from git")
        git.cmd.Git(os.path.dirname(__file__)).pull()
        log.info(f'Updating the mlvm conda env')
        Conda().update_env()
        self.setup_cli()

    def start(self, wait):
        self.set_ml_client()
        poller = self.ml_client.compute.begin_start(self.azureml_config['vm_name'])
        log.info(f"Compute instance {self.azureml_config['vm_name']} is starting")
        if wait:
            poller.wait()
            log.info(f"Compute instance {self.azureml_config['vm_name']} has started")

    def stop(self, wait):
        self.set_ml_client()
        poller = self.ml_client.compute.begin_stop(self.azureml_config['vm_name'])
        log.info(f"Compute instance {self.azureml_config['vm_name']} is stopping")
        if wait:
            poller.wait()
            log.info(f"Compute instance {self.azureml_config['vm_name']} has stopped")

    def ssh(self):
        log.info("Starting ssh client", verbose=True)
        ssh_client = paramiko.SSHClient()
        ssh_client.load_host_keys(const.SSH_KNOWN_HOSTS)
        log.info(f"Connecting to {self.azureml_config['ssh_user']}@{self.azureml_config['vm_public_ip']}:{const.SSH_PORT}", verbose=True)
        try:
            # this is just to make sure we can connect. It's very difficult for paramiko to open an interactive shell session
            ssh_client.connect(self.azureml_config['vm_public_ip'], port=const.SSH_PORT, username=self.azureml_config['ssh_user'], 
                            key_filename=self.azureml_config['private_ssh_key_file'], timeout=4)
            ssh_client.close()
            log.info("Connected! Invoking shell", verbose=True)
            os.system(f"ssh -i {self.azureml_config['private_ssh_key_file']} {self.azureml_config['ssh_user']}@{self.azureml_config['vm_public_ip']} -p {self.const.SSH_PORT}")
        except socket.timeout:
            log.warn(f"SSH connection to {self.azureml_config['vm_public_ip']} timed out.")
            self.set_ml_client()
            if not self.vm_running():
                log.warn(f"VM {self.azureml_config['vm_name']} is not running.")
                if log.yn("Start it up?"):
                    self.start(True)
                    self.ssh()
                else:
                    log.warn(f"Aborted.")
        except KeyError:
            log.error(f"Make sure your {const.AZUREML_CONFIG_FILE} exists and has the required items set")
        except Exception as ex:
            log.error(f"""{ex}
                    Could not connect via SSH.
                    Check your ssh-public-key and the VM's ssh settings in the Azure portal.
                    Azure ML Compute portal: {self.get_ml_computes_url()}""")


    ## STATIC METHODS

    @staticmethod
    def validate_vm_name(vm_name):
        log.info(f"Validating supplied vm name {vm_name}", verbose=True)
        if len(vm_name) < 2 or len(vm_name) > 16:
            log.info("Supplied name is not the right length", verbose=True)
            return False
        valid_chars = set(string.ascii_lowercase + string.ascii_uppercase + string.digits + '-')
        valid = set(vm_name) <= valid_chars
        if not valid:
            log.info("Supplied name has invalid characters", verbose=True)
        return valid

    @staticmethod
    def validate_vm_size(vm_size):
        return vm_size in const.VALID_VM_SIZES
    
    @staticmethod
    def create_config_file_if_not_exists():
        if not os.path.isfile(const.AZUREML_CONFIG_FILE):
            log.warn(f"{const.AZUREML_CONFIG_FILE} file not found. Generating one.", verbose=True)
            with open(const.AZUREML_CONFIG_FILE, "w") as fo:
                fo.write(json.dumps({}))

    @staticmethod
    def get_env_var(env_var, default):
        log.info(f"Looking for environment variable {env_var}", verbose=True)
        try:
            return os.environ[env_var]
        except KeyError:
            log.info(f"{env_var} not found. Using {default}", verbose=True)
            return default

    @staticmethod
    def enter_vm_name():
        while True:
            vm_name = log.usr_in("Enter name for your vm. Recommended format: ds-vm-<yourname> :")
            if not __class__.validate_vm_name(vm_name):
                log.warn("Invalid vm name. Must be between 2-16 chars and contain only letters, numbers, or -")
                continue
            return vm_name

    @staticmethod
    def load_azureml_config():
        log.info(f"Loading {const.AZUREML_CONFIG_FILE}", verbose=True)
        try:
            with open(const.AZUREML_CONFIG_FILE, 'r') as azureml_config:
                return json.load(azureml_config)
        except FileNotFoundError:
            log.error(f"{const.AZUREML_CONFIG_FILE} not found. Please run `mlvm setup` to create it.")

    @staticmethod
    def authenticate():
        log.info("Authenticating...", verbose=True)
        try:
            credential = DefaultAzureCredential()
            # Check if given credential can get token successfully.
            credential.get_token("https://management.azure.com/.default")
        except Exception as ex:
            log.warn("Could not get token. We have to log in to Azure interactively in the browser", verbose=True)
            # Fall back to InteractiveBrowserCredential in case DefaultAzureCredential not work
            # This will open a browser page for
            credential = InteractiveBrowserCredential()
        return credential
    
    @staticmethod
    def setup_cli():
        log.info("Generating the cli executable", verbose=True)
        try:
            PyInstaller.__main__.run(['main.py', '--clean', '--name', const.EXE_NAME]) # generate the executable
        except PyInstaller.isolated._parent.SubprocessDiedError:
            # something weird happened. Just delete everything in build/mlvm and dist/mlvm first
            shutil.rmtree(os.path.join('build', 'mlvm'))
            shutil.rmtree(os.path.join('dist', 'mlvm'))
            PyInstaller.__main__.run(['main.py', '--clean', '--name', const.EXE_NAME]) # generate the executable
        # path to the newly generated executable (src)
        path_to_exe = os.path.join(os.path.dirname(__file__), 'dist', const.EXE_NAME, const.EXE_NAME)
        if os.name == 'nt': # Windows
            path_to_exe += '.exe'
        log.info(f"Path to the executable: {path_to_exe}", verbose=True)
        # path to the soft link (dst)
        linked_file = __class__.select_loc_for_linked_file(const.EXE_NAME)
        log.info(f"Path to the soft link: {linked_file}", verbose=True)
        # Assuming the linked file is in your PATH, this cli will be callable as mlvm
        log.info('Linking files', verbose=True)
        try:
            os.symlink(path_to_exe, linked_file)
        except FileExistsError:
            log.warn(f"{linked_file} already exists. Relinking.", verbose=True)
            os.remove(linked_file)
            os.symlink(path_to_exe, linked_file)
        log.info("Assuming the linked file is in your PATH, this cli will be callable as mlvm")

    @staticmethod
    def select_loc_for_linked_file():
        log.info("Finding a location for the linked file", verbose=True)
        # select the env dir that contains the exe_name, which is assumed to be the same as the env_name
        env_dirs = json.loads(subprocess.check_output('conda env list -v --json'.split(' '), text=True))
        for env_dir in env_dirs['envs']:
            if env_dir.endswith(const.EXE_NAME):
                return os.path.join(env_dir, const.EXE_NAME)
        log.error('Could not find conda env mlvm. Please find the mlvm executable and link it to your PATH manually.')


    ## OTHER METHODS

    # this ensures that self and the config file match. Use this when adding/updating config attributes
    def add_item_to_azureml_config(self, key, value, allow_none=False, prompt_overwrite=False):
        if not allow_none and value is None:
            log.error(f"Can't accept value of None for {key}")
        self.azureml_config = __class__.load_azureml_config()
        existing_value = self.azureml_config.get(key)
        # see if the user wants to overwrite a new value into their config
        # only necessary if the existing value is not none; !still necessary if the new value is None!
        if prompt_overwrite and existing_value and existing_value != value:
            if not log.yn(f"Overwrite {{\"{key}\": \"{existing_value}\"}} with {{\"{key}\": \"{value}\"}}?"):
                log.info(f"keeping {key}: {existing_value} in {const.AZUREML_CONFIG_FILE}", verbose=True)
                return existing_value
        elif existing_value == value: # If they're already the same then who cares
            return value
        log.info(f"Adding {key}: {value} to {const.AZUREML_CONFIG_FILE}", verbose=True)
        self.azureml_config[key] = value
        with open(const.AZUREML_CONFIG_FILE, 'w') as file:
            file.write(json.dumps(self.azureml_config))
        return value

    # try to use the passed var first
    # if that's not set then try the os.environ
    # else use the value in const.py and set to os.environ
    def set_ml_env_vars(self, resource_group, subscription_id):
        if resource_group is None:
            self.add_item_to_azureml_config('resource_group', __class__.get_env_var('RESOURCE_GROUP_NAME', const.DEFAULT_RESOURCE_GROUP), prompt_overwrite=True)
        if subscription_id is None:
            self.add_item_to_azureml_config('subscription_id', __class__.get_env_var('AZURE_SUBSCRIPTION_ID', const.DEFAULT_SUBSCRIPTION_ID), prompt_overwrite=True)

    def initialize_azureml_config(self, workspace, resource_group, subscription_id, vm_name, vm_size, ssh_user, ssh_public_key_file, ssh_private_key_file):
        self.set_ml_env_vars(resource_group, subscription_id)
        self.add_item_to_azureml_config('workspace', workspace, prompt_overwrite=True)
        self.add_item_to_azureml_config('ssh_user', ssh_user, prompt_overwrite=True)
        self.add_item_to_azureml_config('vm_name', vm_name, allow_none=True, prompt_overwrite=True)
        self.add_item_to_azureml_config('vm_size', vm_size, prompt_overwrite=True)
        self.add_item_to_azureml_config('ssh_public_key_file', ssh_public_key_file, prompt_overwrite=True)
        self.add_item_to_azureml_config('ssh_private_key_file', ssh_private_key_file, prompt_overwrite=True)

    # const.AZUREML_CONFIG_FILE should be in json format with keys workspace_name, resource_group, and subscription_id
    def set_ml_client(self):
        if self.ml_client:
            log.warn("ML Client is already set. Not recreating.", verbose=True)
        log.info("Getting ML Client", verbose=True)
        credential = __class__.authenticate()
        self.ml_client = MLClient.from_config(credential=credential, path=const.AZUREML_CONFIG_FILE)

    def setup_ssh(self):
        log.info("Setting up SSH", verbose=True)
        # save new/existing ssh key file paths in azureml_config
        log.info("Saving the new SSH key file paths in azureml_config")
        self.link_ssh_keypair()
        return ComputeInstanceSshSettings(ssh_public_access='Enabled', admin_public_key=self.azureml_config['public_key_file'],
                                          admin_user_name=self.azureml_config['ssh_user'], ssh_port=const.SSH_PORT)

    def link_ssh_keypair(self):
        if not os.path.isfile(self.azureml_config['public_key_file']):
            log.warn(f"Could not find existing public key file {self.azureml_config['public_key_file']}")
            if log.yn(f"Generate new ssh keypair with private key at {self.azureml_config['private_key_file']} \
                      and public key at {self.azureml_config['public_key_file']}?"):
                # generate a new ssh keypair if we don't have one
                # Assumes we don't have one if the specified public_key_file doesn't exist
                self.generate_ssh_keypair()
            else:
                log.error("Did not link ssh keypair. To link an ssh keypair, please specify the file locations for existing ones, \
                          use the default locations, or generate new keys. You can specify specific key locations with the \
                          --ssh-public-key-file and --ssh-private-key-file options in `mlvm setup`\n \
                          Aborting.")
        else: # assumes we already have an ssh keypair so we just need to note them in the azureml config
            # except, we should already have those file locations set by this point, so we don't need to do anything
            pass

    def generate_ssh_keypair(self):
        log.info("Generating SSH keypair", verbose=True)
        private_key = crypto.asymmetric.rsa.generate_private_key(
            public_exponent=const.RSA_PUBLIC_EXP,
            key_size=const.RSA_KEY_SIZE
        )
        public_key = private_key.public_key()
        if not os.path.exists(const.SSH_DIR):
            os.makedirs(const.SSH_DIR)
        log.info("Saving SSH keys", verbose=True)
        self.save_private_ssh_key(private_key)
        self.save_public_ssh_key(public_key)

    def save_private_ssh_key(self, private_key):
        private_pem = private_key.private_bytes(
            encoding=crypto.serialization.Encoding.PEM,
            format=crypto.serialization.PrivateFormat.PKCS8,
            encryption_algorithm=crypto.serialization.NoEncryption()
        )
        with open(self.azureml_config['private_ssh_key_file'], 'wb') as f:
            f.write(private_pem)

    def save_public_ssh_key(self, public_key):
        public_pem = public_key.public_bytes(
            encoding=crypto.serialization.Encoding.PEM,
            format=crypto.serialization.PublicFormat.SubjectPublicKeyInfo
        )
        with open(self.azureml_config['public_ssh_key_file'], 'wb') as f:
            f.write(public_pem)

    def load_public_ssh_key(self):
        log.info("Loading public ssh key", verbose=True)
        with open(self.azureml_config['public_ssh_key_file'], 'rb') as key_file:
            try:
                return crypto.serialization.load_pem_public_key(key_file.read())
            except ValueError as ex:
                log.error(f"""{ex} Your key needs to be in the PEM format. 
                        Please look up how to encode it as such (recommended) or generate new ssh keys.""")

    def add_to_known_hosts(public_ip):
        transport = paramiko.Transport(public_ip)
        transport.connect()
        key = transport.get_remote_server_key()
        transport.close()
        hostfile = paramiko.HostKeys(filename=const.SSH_KNOWN_HOSTS)
        hostfile.add(hostname=public_ip, key=key, keytype=key.get_name())
        hostfile.save(filename=const.SSH_KNOWN_HOSTS)

    def link_vm(self, ssh_settings):
        log.info("Linking vm", verbose=True)
        if self.azureml_config.get('vm_name') is None: # prefer .get() to [] here because we're filling it in here anyway
            try: # look for vm_name in the configs
                if not log.yn(f"Found vm_name {self.azureml_config['vm_name']} in {const.AZUREML_CONFIG_FILE}. Link this one?"):
                    self.add_item_to_azureml_config('vm_name', __class__.enter_vm_name())
            except KeyError:
                self.add_item_to_azureml_config('vm_name', __class__.enter_vm_name())
        # provision vm if doesn't exist
        # if successful, self.compute_instance will be populated
        self.provision_vm(ssh_settings)
        if not self.compute_instance:
            log.warn("Did not provision VM")
            return False
        self.add_item_to_azureml_config('vm_public_ip', self.compute_instance.network_settings.public_ip_address)
        self.add_to_known_hosts(self.azureml_config['vm_public_ip'])
        return True

    def provision_vm(self, ssh_settings):
        try:
            self.compute_instance = self.ml_client.compute.get(self.azureml_config['vm_name'])
            log.info(f"Found existing VM {self.azureml_config['vm_name']} in {self.ml_client.workspace_name}")
            log.info(f"If this is yours, feel free to continue with setup to link it.")
            yn = log.yn("Would you like to continue?")
            return yn # True if we want to continue because we have self.compute_instance set
        except ResourceNotFoundError:
            yn = log.yn(f"Could not find existing VM {self.azureml_config['vm_name']}. Create?")
            if not yn:
                return False
            log.info(f"Provisioning VM {self.azureml_config['vm_name']}. This may take a few minutes.")
            self.compute_instance = ComputeInstance(
                name=self.azureml_config['vm_name'], size=self.azureml_config['vm_size'], ssh_public_access_enabled=True,
                ssh_settings=ssh_settings, idle_time_before_shutdown_minutes=const.IDLE_TIME_BEFORE_SHUTDOWN_MINUTES,
                setup_scripts=None, enable_node_public_ip=True
            )
            self.compute_instance = self.ml_client.compute.begin_create_or_update(self.compute_instance).result()
            return True

    def vm_running(self):
        self.compute_instance = self.ml_client.compute.get(self.azureml_config['vm_name'])
        return 'running' == self.compute_instance.state

    def get_ml_computes_url(self):
        url = 'https://ml.azure.com/compute/list/instances?wsid=/subscriptions/'
        url += self.azureml_config['subscription_id'] + '/resourceGroups/'
        url += self.azureml_config['resource_group'] + '/providers/Microsoft.MachineLearningServices/workspaces/'
        url += self.azureml_config['workspace_name'] + '&tid=1aff0669-ee5f-40b8-9800-b5ec4f39c48e'
        return url

    def setup_sf_connection():
        log.warn("Snowflake Connection setup has not been implemented yet.")
        pass
